src/JsonTxBuilderBase.sol

```diff
@@ -51,4 +52,95 @@ abstract contract JsonTxBuilderBase is CommonBase {
 
         return calls;
     }
+
+    /// @notice Reads all account and storage accesses and makes a series of basic sanity checks on them.
+    ///         This function can be overridden to provide more specific checks, but should still call
+    ///         super.checkStateDiff().
+    function checkStateDiff(Vm.AccountAccess[] memory accountAccesses) internal view virtual {
+        require(accountAccesses.length > 0, "No account accesses");
+
+        for (uint256 i; i < accountAccesses.length; i++) {
+            Vm.AccountAccess memory accountAccess = accountAccesses[i];
+            require(
+                accountAccess.account.code.length != 0,
+                string.concat("Account has no code: ", vm.toString(accountAccess.account))
+            );
+            require(
+                accountAccess.oldBalance == accountAccess.account.balance,
+                string.concat("Unexpected balance change: ", vm.toString(accountAccess.account))
+            );
+            require(
+                accountAccess.kind != VmSafe.AccountAccessKind.SelfDestruct,
+                string.concat("Self-destructed account: ", vm.toString(accountAccess.account))
+            );
+
+            for (uint256 j; j < accountAccess.storageAccesses.length; j++) {
+                Vm.StorageAccess memory storageAccess = accountAccess.storageAccesses[j];
+
+                if (!storageAccess.isWrite) continue; // Skip SLOADs.
+
+                uint256 value = uint256(storageAccess.newValue);
+                if (isLikelyAddressThatShouldHaveCode(value)) {
```

Comment:
Now that I think of it, we can make this check even stricter—we currently have a no-op if `isLikelyAddressThatShouldHaveCode(value)` returns false. Instead we can assert that there is no code, as we expect.